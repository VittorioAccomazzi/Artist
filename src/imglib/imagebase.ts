import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}