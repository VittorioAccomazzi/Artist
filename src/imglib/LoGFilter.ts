import { Image2D, ImageFactory, ImageFloat32 } from './imagebase';


export default class LoGFilter {
    static Run( inImg : Image2D, sigma : number ) : ImageFloat32 {
        let kernel= this.generateKernel(sigma)
        return LoGFilter.applyKernel(inImg, kernel);
    }

    private static applyKernel(inImg: Image2D, kernel: number[][]) : ImageFloat32 {
        let width = inImg.width
        let height= inImg.height
        let ouImg = ImageFactory.Float32(width,height)
        let hKSize= (kernel.length-1)/2
        for (let y = 0; y < height; y++) {
            for (let x = 0; x < width; x++) {
                let wSum = 0;
                let vSum = 0;
                for (let i = -hKSize; i <= hKSize; i++) {
                    let y0 = y + i;
                    if (y0 >= 0 && y0 < height) {
                        for (let j = -hKSize; j <= hKSize; j++) {
                            let x0 = x + j;
                            if (x0 >= 0 && x0 < width) {
                                let weight = kernel[i + hKSize][j + hKSize];
                                wSum += weight;
                                vSum += inImg.get(x0, y0) * weight;
                            }
                        }
                    }
                }
                if ( Math.abs(wSum) > 0)  ouImg.set(x, y, vSum / wSum);
            }
        }
        return ouImg;
    }

    private static generateKernel( sigma : number ) : number [][] {
        let hKernel = Math.floor(4*sigma)
        let kernel = Array(2*hKernel+1).fill([]).map((v,y)=>{
            let y0=Math.abs(y-hKernel-1)
            return Array(2*hKernel+1).fill(0).map((v,x)=>{
                let x0= Math.abs(x-hKernel-1)
                let sigma2 =sigma*sigma
                let val = (y0*y0+x0*x0)/(2*sigma2)
                let weight = -(1/(Math.PI*sigma2*sigma2))*(1-val)*Math.exp(-val)
                return weight
            })
        })
        return kernel
    }

}