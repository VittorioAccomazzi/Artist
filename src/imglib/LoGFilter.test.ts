import { getCanvases, hash, dumpCanvas, toSeqCanvas, dumpImage, toNodeCanvas } from './testutils'
import CanvasUtils from './canvasUtils'
import { ImageFactory } from './imagebase'
import LoGFilter from './LoGFilter'

test('shall detect edges', async ()=>{
    let width = 200
    let height= 280
    let bck = 50
    let frg = 100
    let image = ImageFactory.Uint8(width,height)
    let pixels= image.imagePixels
    pixels.forEach((v,i)=>{
        let y = (i/width | 0) 
        let x = (i%width)
        let val = bck
        if( Math.abs(x-y)<8 ){
            val=frg+(x | 0)
        }
        pixels[i]=val
    })

    //await dumpImage(image, 'LoG input image')
    let logImage = LoGFilter.Run(image, 2)
    //await dumpImage(logImage, 'LoG  image')
    let hsh = await hash(logImage)
    expect(hsh).toMatchSnapshot();

})

test('shall enhance edges',async()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        let [lImg, aImg, bImg ] = CanvasUtils.toLab(toSeqCanvas(canvas))
        
        let logImg = LoGFilter.Run(lImg, 2)
        // await dumpImage( logImg, `loG image ${name}`)
        let hsh = await hash(logImg)
        expect(hsh).toMatchSnapshot()

        let lPixels = logImg.imagePixels
        let scale = 0.001
        lPixels.forEach((v,i)=>lImg.imagePixels[i]+= scale * v)

        let sCanvas = CanvasUtils.fromLab(lImg, aImg, bImg)
        // dumpCanvas(toNodeCanvas(sCanvas), `enhannced ${name}`)
        hsh = await hash(toNodeCanvas(sCanvas))
        expect(hsh).toMatchSnapshot()
    }
},60*2*1000)