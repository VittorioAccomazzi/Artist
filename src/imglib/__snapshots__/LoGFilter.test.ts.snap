// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall detect edges 1`] = `"000000000000000000000000000000001e0000000000000000000000000000003f8000000000000000000000000000003fc000000000000000000000000000003fe000000000000000000000000000001ff800000000000000000000000000000ffc000000000000000000000000000003fe000000000000000000000000000001ff8000000000000000000000000000007fc000000000000000000000000000003ff000000000000000000000000000001ff8000000000000000000000000000007fc000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000001ff000000000000000000000000000000ffc000000000000000000000000000007fe000000000000000000000000000001ff000000000000000000000000000000ffc000000000000000000000000000003fe000000000000000000000000000001ff800000000000000000000000000000ffc000000000000000000000000000003fe000000000000000000000000000001ff8000000000000000000000000000007fc000000000000000000000000000003ff000000000000000000000000000001ff8000000000000000000000000000007fc000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000001ff000000000000000000000000000000ffc000000000000000000000000000007fe000000000000000000000000000001ff000000000000000000000000000000ffc000000000000000000000000000007fe000000000000000000000000000001ff800000000000000000000000000000ffc000000000000000000000000000003fe000000000000000000000000000001ff800000000000000000000000000000ffc000000000000000000000000000003ff000000000000000000000000000001ff8000000000000000000000000000007fc000000000000000000000000000003ff000000000000000000000000000001ff8000000000000000000000000000007fe000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000003ff000000000000000000000000000000ff8000000000000000000000000000007fe000000000000000000000000000001ff000000000000000000000000000000ffc000000000000000000000000000003fe000000000000000000000000000001ff000000000000000000000000000000ffc100000000000000000000000000003fe180000000000000000000000000001fe3e00000000000000000000000000007e1f80000000000000000000000000003e0fc0000000000000000000000000001e07e0000000000000000000000000000f03f0000000000000000000000000000700fc0000000000000000000000000002007f0000000000000000000000000000003f8000000000000000000000000003c00fc000000000000000000000000007f007f000000000000000000000000001f801f800000000000000000000000000fc00fc000000000000000000000000003f007e000000000000000000000000001f801f000000000000000000000000000fe00f0000000000000000000000000003f0030000000000000000000000000001f8000000000000000000000000000000fe0000000000000000000000000000003f0000000000000000000000000000000fc0000000000000000000000000000007e0000000000000000000000000000003f0000000000000000000000000000001fc0000000000000000000000000000003e0000000000000000000000000000001f0000000000000000000000000000000f000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 1`] = `"00000000000000000000000000000000023101473d600004030690080288008806860e204569f1c140200dc2361260005050f38888be8e027aabd215c96b477452adac6eca005cffedd9fdcc4645dea0266e67901504108010ea63a3a1a99c3020601078efea805e2d2141a4305770ca2482490fdb77ffb7c2f0245046a88bb43a3826407c8c0d9dde93e78f402a80840e62cc584002313468cffe36d549e8f0519c7febbdf7cfe2f392086a1623274e3dc2a654496044df9f6f5dbb48b4e256404f44aa10980031600b2604a64d280848baa5426e6ed6c6613051a2cc83968038c42a2c887b6ffbff8f15dacd7670b46f8e6bb8c34545efbc63cf6b09aa8ea26272d567545a52d08794e016be14dcd82ccd80a2299c0d5c74376aaf03d95252422a2b524beffbe3ffe0efbec9964fa60d4ce95a3afe208edb9f9e53b323a0d06cbbd7ed57905056022ff4c01c4c134c5ca3017cc003aff84740333d7d55ff322add3ae0b0fefebfebb720d180dbae40327a471a2fbf4a97ff747d472c1153f43f0abbb9fb41bb580c1fd3fc34ef28161bbe6df3aaf207efc9e2aea7671d2f2e49f3a75f55dfffbab24adab402e9d250581694a275218945bf7c1d547eeb2d422795e9bb5fdf6bddfa97e6c1e7f5e6385ebe846dcb723afb3f58db00ec0ac194566cdb9604ddcfbad1ca707f042474d467be67fbfffcffaffd7fb07e7c7fcf121801884d2882200012605ff88ff228880050437006ad569fae90bfe8410053720bffa8569104a800c02043f15427805634c85ba4d8723fffdaeb010f48ca43905322e6ee87bde5ae397fdcffb630210c28841817b04afa2017403fc02a433c524b2573f80bfa459bea3e7f800120a11023bd5743ef0db1e42d6fe002de9dc39c7e83c07fe175b67ba90e001e1b591400217ea2f98ed51eb56240bfe8440aa25c1d1028012933f4553689015621819422130ccbf3226ad3284d7642400c82612017f21f05f8a60bffb6134febfafe0c1e6cdde033129b75b2ef46200853c947000a0003edef5295749d7c54096406ac0c5fed960b29eefb59d7137bfaebfa57e40147fadc6a1bc37f2d34b565ac3408c460410242f5eff151ebf572af557bc450062c5e511f14176412166c00d2015e62156ba1dd399731b5edeffbf21e468ca456ab3429fe67088d1fb50407a1bdd3501a88d7dee1f8ff7757b575f63bf7b18c5a531d779eeffffee0496486497e86ba402061ca53bfec1fff7e8778308db5cc09ffcff5d63e90c02ff7bee78cfcffff330210c277e0155ff97dbc2adfffffff0413025df816fffebfafffffff0364002fd88352032dec00047ee7dff00109f07a914ac1ffff1ffffffa400000cc851f0e2f75bfebbfffffffb008100ae7807737e8ab7efffdfd0000046f5fb86af885bfd7fffffff8d0007fa2b26584040000425fe7ffffff37ff9fa6784005ffff1a7ffffffffff5804074721109ff900ffffffffffffff4015c050f8ffffffffff8ffffffffc007ff8180007a3000000000fffc00000000000000000000000021e40000003ffffffc0ae02cfb500083980000ffff903c0017fffffffffff01dfffe4e20000383bc000000002f2ffffffffeffdbffb77023f277ffff06193b208240e001ccb887ff9fcdc0a0008406ae3fe81bbfbff57f0058f0100fdfffa9902c1c43d0000100f3a615f751d0873443d1e60006c8c07f3dd9e6d1fd2778e5548832fffc4c58bec65602f83c489e3ffe3fdef001f3ffc1c5f43e05c3bfe74083e77c0379ff80631bd394abf4001020000000fcff2c5fdef02fcc103feffd6fff0afe1fd04308019fd43befc03c0008e19f9ee002887ffa20edc0007c004ee4000040ffba27fe83780017ef8ffff78f4a083e0003fe00fc0fffec906080007fff1f886bfc03f003f0000b74077c4801c0f5fe58e2c01fe002ff0003ff8f760e1d8000ee003ffc007f001fffc40001effc3dde1fffc000ffc007e700007fffc227c000e78000380003fff03ff385c014583ffe201fffc0ffff000fff8ff001a2a1c0082f81c07ffc009fffe07003feffffffe2fffc0f78000c30001c003ec0000020180007fd813fffffffc3fffffffef600de3dfa0026fff380000e000004010ffce2c7819fdff00000fff1ffffebc0f04398bc0000f8017ffff87f038ff81f073c07007fff8ff001f80000000027fffff2103ff040300ffc07fffffc7000003800fff33f861ffc3fffe0003fffffe703ffc080007f800080000fffce000fd9d30000f83180e0fffbfff8000066003f7cffff3f8e7f9fc3ff1fe7fffffefff1f98017006390e7ff000000000007800067ff38fff1fffc00ffffffffff40000fcc0fc73f8f80019f0603fe0025fffffc73f83cc07807fe70600820e183800e03ce200307f0bfc3e7ffffff9e3c0fe0e4000e60f83fc00f1c00000185fff84e100fffffffc03fff2000ffe0fffc131f2fffe000001e00001fffff1c0003ede1781e3ffeffe3ffe7e1fc01e3e08c003ff02000f0800004181000e03fff7e7fe00fefffffffffe300ef07fc80007b800000180800b6603cffd9fe07ffe7e03ecff9800040ffdfc30000001c01fe1fc33b0ff9ff3f00000c081fc7f93000c03cf0ffeffff00738207f80018f8001fde20019bfc01ff8007f80f3fc603efe07fffffff87e00e3ff83fe3f878c1fc3f8380f800007801cf07e33e00001e00047c20070d1fff7e79f908000c01e7f3c3020fe0f00015c8000687c1023f601d7cfdf71f3ff0d802d9f8791c7fd39fe7403f0df1818f078fff078effc10000000b00fce0fe01dcf0101f3fcff0378001e1fc00080ffc1b80070e4c1ec68fe0a87c1ffe83816008080000000000000000000000000000000"`;

exports[`shall enhance edges 2`] = `"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000010f9d000000000000000000000000005fffffff000000000000000000000004fffffffdf80000000000000000000003fffffffffd200000000000000000007fffffffffff80000000000000000007fffffffffffffb00000000000000003ffffffffffffffff0000000000000017fffffffffffffffff80000000000005ffffffffffffffffffe004000000003ffffffffffffffffffffe0800000000ffffffffffffffffffffff7000000000ffffffffffffffffffffffe40000007bfffffffffffffffffffffff7c00000fffffffffffffffffffffffffffe600bfffffffffffffffffffffffffffd700ffffffffffffffffffffffffffffff90ffffffffffffffffffffffffffffffa0ffffffffffffffffffffffffffffffe0fffffffffffffffffffffffffffffff2ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000020000000000000000000000000000003e0000000000000000000002b79000001c000000000000000000001fffffa0000000000000000000000001ffffffe1f00000000000000000000007fffffffc41800000000000000000003ffffffffff5c0000000000000000000ffffffffffff00000000000000000007fffffffffff80000000000000000001ffffffffffff80000000000000000001ffffffffffffc0000000000000000003fffffffffffff000000000000000001ffffffffffffffc0000000000000000fffffffffffffffe0000000000000001ffffffffffffffffc000000000000003fffffffffffffffff80000000000000fffffffffffffffffff8000000000003fffffffffffffffffffe000000000007ffffffffffffffffffff00000000001fffffffffffffffffffffe0000000003ffffffffffffffffffffff800000000ffffffffffffffffffffffff00000003ffffffffffffffffffffffff80000007ffffffffffffffffffffffffc000001ffffffffffffffffffffffffffc0000fffffffffffffffffffffffffffffda1ffffffffffffffffffff7fffd002000fffffffffffffffffffe620130000000fffffffffffffffffffa000000000000ffffffffffffffffff80000000000000fffffffffffffc000000000000000000fffffffffff8d0000000000000000000fffffffffffffbffffa0000000000000fffffffffffc0070f070000006000000fffffffffffc005e6048000000000000ffffffffff0000000000000000000000ffffe000000000000000000000000000000000000000000000000000000000000000000000000003fffffffe000c000000000000000002bff77fffffffffffff0000000010effffffffffffffffffffb000000003fffffffffffffeffffffffb00000017ffffffffffffffffffffff7f00000007fffffffffffffffffc53ffff0000006ffffffffffffffffffffffffb121e1ffffffffffffffffffffffe3ffe7017fffffffffffffffffffffc00800183fffffffffffffffffffdf800000000fffffffffffffffffffffffd00000000fffffffffffffffffffffe0000000000ffffffffffffffffffffc20000000000fffffffffffffffffffffff380000000ffffffffffffffffffffc00000000000ffffffffe7ff07ffffffff4000000000ffffffffff9afffe03ffa00000000000ffcfffffec7ff81fffdc0000000000009fffff00fff3dfff0000304000000000e7e37e7f783ffd700020000000000000101bffc1ffff02073f820000000000000500e37fff0000fdf000000000000000c88c06780004f24000000000000000000003ff801fffffc4003fc0000000000000000006fff00000060000000000000003801c0ff80000018010000000000000fffffffffffffffffffffffffffffffffffffffffffffffff63ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff9fffffffffffffcffffffffffffffffeffffffffffffffffffffffff07fffffffffffffffffffffffffffffffffffffffffffffffffe4441cfefffffbfffffff9fffffffff0ffffffffffff7cbffffffffffff801ffdffffe077ffffffffffffffffffffffff000700187fefffffffe3fffffffcfffffffffde7f8ff1e7e1dff9ffffffffdf800003efffffeff90ffffffff73ffc00004000ffff101ffffffe07301e05941fffff0c00006de9fc1ffffffdfffb8047ff01e3c000003ff02001f0f370040c1000007dfe00010003c7fffffffe2000320300000010000000080000fff8bffffdfc03ff01000000d0000000ffffc10000000000400000010fb8031f0000000003007800000018c07ffffff000000000000180000000c00011fff8074800000061fc0000300000fd06004000a30000ef9f82000000000000000000001ef80003e00000000000000000000000e780000000000000000000000000000080000000000000000000000000000000200f8300000000000000000000000000fff8006300000000000000000000000f000000180000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall enhance edges 3`] = `"00000000000000000000000007800002287e300c08480c8180c47c0227cc987c2c7e245b585b000868543cd2a3c7029800fe2d1718b2dedc297d3ef493e4c27019fe2d39fe9f5edfecdd1f76d1e3e4e079ae382ae8b011de78161fb811f107c061ce32599613c303cbb61f8b2cf9e180238e3c1da6e9e148ce3a1fbd7c79e70c378e3a3ff92c818eb8be1f97be7cee382f9e3038518c000f885d9fdcfe3c5c3c1d4e3a0b26208003ad749d48703e1868185e3c6a653803f07e7c1c73e39e30f038ce39698fd8fffe1ef29c77d01f20f078de3b3dfb0bc5c787559cb2108f878c79fe38ae0c7bb22fe3701e98400782d479fe183ad9e30e3871fe1cdc6de78730f1ee3288eb0320ba7c7f1e2c24430f26f3de3c38325b43c37e1d9fb6291206cc63ee3405160dd01ee31d9fc76bdc188863fe101c442f6479c7cd9fc72e1c3911c3fe2811c8d8407f1ee61ce7a358733703fe2b3c3971617cf8739de39430e66613fe362db0c68e33ecb19ff3d775ec0c33fe305c6586b8c6e7b89ef3e061c81831fe3ccd49e4339f085e1ff1f0c3923a71fe32be4c0bea31c7af1ef0f3c6b46653ce3e1e8c2e78fc7aef1ff0e1872cce71fe1b7c81ac71d9b7079ff0190859dc01ef364d8cbec763bc359ef23f1b919811ee280197398f1b63d8cff0ef14b33451fe1a0994631c0ecf08cff1bb19267c19fe3a1918ce72cd35fc67e3611a66c029fe3d69171ce9b67ac6f3cee330cdf831de3f3b1d7185c9db9f3a1b871d998041fe3c1b12434733a73f98670f281b9c01fe3a3310a69a64888cc9cc6730339041fe241307ce399a3bb94738ef12370041fe3832249a67358e7188e9ce1266b831fe2d32213ce266536c33d1de304e9401ee3832f269b8d8c584ce11a660dd8c01ee3c7610d38b3f54f1383324499c4001fe3a1751e737ee7f66e17764d1b88c41fe3137534e6c21ffc984e6ccd3308c01fe31773b9cd98cfe773fdf18936a8401fe35379b39360e004ecefc39266c2441fe3973827365e451b8dae07326df1841de3d439ce6cf0f937e8780e24dc04001ee315399cdd919d6cc0e07ce4d812c21ce379bcb98f2130d94b87f9c99e90801ee3119b3234d271331f177049be92061ce3c1996671962af25a1bc1813d87801fe2d9ce2e7b36adf6432b84837106011ef281ee19c629a1a11f0705b27368829ef0e0ef576290174c73870b06f9c9861fe3e0ee6f03982fccf3830b26bc68809ff3a4661ba18a5b99f783222d8f61839ff0f1a4b86c28b133e7c3364d81ffc11ff3c1767c2a3d7b45f7c32e49c03f011fe2b172f12987fb0dffc3349bf003c29fe1b080f519359b48c3c12c9b1d20c01fe381df6931659cfc6fe12d939f80019fe0d1998b03cbb81e47c12836067c039fe2c0119351df37866e31b937423e079ef3887f15589f6df27df9992c462fc39ff0c9fe654c3bc0187811916e762cc01ef00338cd467381887191926c14e9431ef31b395c607708c06181a05a010a029ff3b0e21d64aec9c06981a6d80bb8029ff3e786d41cfc67e0efb1a0d86ca0e69ff3ef1cde09597824602484b60601e29ff1b80e2413f3e1e4676485b4cd47c09ff30330d412e6ffec67e481b8538f019ff07e7c7f05ccbec869e48162c01e019ff1e0b83325cdbfe864e4c9684a3c019ff04da1030b9cbecc66c4cb483670629ef27733238b1ecec422c4c3c00c70e39ef2ccba32d70e77e470ccc2f844e1c09ef358ba37164e53e037ccc6ce14c7c31ff144912c0c4e23a436cce68279ce669ef6cd94346c6ed5af378ce7a1f18c631ef0c9358c786cbbe8378ce5842399621ef3db3cc8726df5e2230c6d8433bdc31cf2d2ecd3b26c25fae30c6d12b333133cf0cc76c7632ee7d3660e6f00d362f13cf1d8168e432e23c3660a6a046668b21cf3d31726912ee3cfe49a6bca66d9b31cf290c93d932715fc8c9e6b0a67dbb21fe486f65911273c00092d7b0a67c0f21eed82d8d253273fffb95d7222e7aff21ef1b6c8d64ba70f2fb15d3394e707825ff3a496acd7b6a30072553240673c0056f224b238d6b72200e195347c6778f0d6f324ba325093f81fc1953701c743e0d6f364b9b2d094ffffd355379e670fe4d6f2649d86d4847fffd114374c6639e07ef26c8e82d442001a991337fc467fc05ee748c8f0d43b0018051b31ff4eff005ecd74e2d8d229ffffe115200fccbc005ed264664c9351e0c1f8352c03ccfc600e66366c569118e603fe156f800cf0668e64373d5a9188d983bf9d6ef04df3c78ecc3c58e5108df38fefcb7ffecfc4c78ed9a99175159ced9db3e37bfedfccc78ed462516a918ccd1d0ff358ecdfd987ced0534859908cfe1488725fea9f9303cee8d62709988ce3b72f38505c1f8663cee8869b19998ceda4d3d85bd8ff2e43cef986c3bd998cec248bdc5e90df1cc3cef30c49999a1cf82329fcf2c1de1ce3cff21ccab7011cfdba3dacecd1de1983ce7836ca73e51cec030dbee0e19c19c3cff1864a783c3cca1526f6f461bc3303cff23666ac047df661e74ef441b9a741ced8e4ad7750f9e80d07fa7fc0396661cef885e9135079cd340fff0fee7a4ce1cff09d098710f9bf8fa37f03fc78d8e1cf759517cc20f33a0439fb003c7091c1cf6435d8f9c1e762438977b806713181cf68d7646d41e78788675f9f06e36381ef6993823c07ced5a424fb9f86e4c341e7e8114e888f9dc2a4987edd87c5c685e6e9d36d351c3fbfb04a3fcca7098704e7e9596531c0fda5fa29becf26130f20e7f05964b907d84cc82aefc330661f00e7e04956911ff996fc400367f8cc1e00e7e28b52583eee24bdb34f66c1dc3c00e670b35a4c700808001213e4a1907c00e060000200000000000000700000001"`;

exports[`shall enhance edges 4`] = `"fffe0000b5ffffffbd807fffffc00003fffe0000adfffcdf90003fffffc00000fffe00000bffeccce0003fffffe00018fffe000019ffffdfe1001fffffe00030fffe00001ffffe5fe4d01ffffff000e0fffe000009fdffcea0001ffffff001c0fffe0000141bc003c0001ffffff80180fffe000003f9c001c0001ffffffc0300fffe000001bdc003c0001ffffffc0600fffe0000000c0001c0001ffffffe0c00fffe00000004000184000ffffffe0800fffe000001bc8ff80c000fffffff0000fffe00000298ffff0e000fffffff0000fffe00000903ffffc1000fffffff8000f9fe0000003bffffe0000fffffffc000f9fe000001f1fffff8000fffffff8000f1fe000001037ffffc000fffffff8000f1fe00000003fffffe000fffffff0000e1fe00000200ffffff800fe7fffc0000e1fe000004017fffffc00fc7fffc0001c1fe000000007fffffe00fe3fff8000381fe00000000fffffff00fe1fff0000781fe00000000fffffff80fe0fff0000f01fe00000001fffffff80fe0ffe0000f01fe00000001fffffffe0fe07fc0001f01fe0000400bffffffff0fe07fc0003f01fe0000c00ffaffffff0fe03f80007f01fe0000800ff1ffffff8fe01f0000ff01fe0000801fc7ffffffcfe03f0001ff01fe0001803f8fffffffcfe07f0001ff01fe000180231fffffffcfe0ff0003ff01fe000180443fffffffe7e3ff8007ff00fe000180007ffffffff7c7ff8007ff00fe00030000ffffffffff9fff000fff00fe00030001ffffffffffffff000fff00fe00030007dfffffffffffff001fff00fe00038007bfffffffffffff003fff00fe0003800f7fffffffffffff003fff00fe0003801ceffffffffffffe007fff00fe0003c029fffffffffffffe007fff00fe0003c053fffffffffffffe00ffff00fe0003c087bffffffffffffc00ffff00fe0003f10a2ff1ffffffffcc01ffff00fe0003f008580cfe7fffff1801ffff01fe0003d8103a0e007ffffc1803ffff00fe0003c0202400007ffff03003ffff00fe0003c8604800017fffc0e007ffff00fe0003d800900002ffff03c007ffff00fe0001e900400005fffc7f800fffff00fe0001f10000001bfff03f000fffff00fe0001f00000000fffe03e000fffff00fe0000f00000000ffff0380007ffff00fe0000f00000001ffff0700007ffff00fe0000700000003ffff8300007ffff00fe0000604000007ffff830000fffff00fe000060000001fffff830001fffff00fe000060000003fffffc30005fffff00fe000060000007fffffc10005fffff00fe00002000000fbffffc1000bfffff00fe00000400000ffffffc1000bfffff00fe00001000001ffffffc10007fffff00fe00018800003fc1fffc10017ffffe00fe00011000007ff87ff018007ffffc00fe0003800000fe083fc00800fffffc00fe000fe00000fc013f800802ffffff00fe001f804003f808bf000803ffffff00fe001f000003f00c3f080807ffffff00fe000e000003e01e1e180807fffff300fe003c004007c67e3e780807ffffff00fe00700040079f977fc00807ffffff00ff01c0c0401f1fffffe0080fffffff00fe010001000e3ffffff0080ffffdff00fe03e000401e1fffeff8080fffffff00ff0f0080001c1fffeffc0c1fffffff00ff00000010381ffffffc0c1ffff7ff00ff01000008303fffdff80c1fffefff00ff04410020703fffcff80c3fffefff00fe1c810030601fffcff80c3fffdfff00fe04010000c01ffe4ff8043fffdfff00fe20010003c01ffef7f8063fffbfff00ff00000003800ffe87f0067fffbfff00ff00108007001fff03f0067fffbfff01ff0000c41b000fffffe0067fffffff01ff04004072000fffffe0067fff7fff01ff08002074000fffffe006ffff7fff01ff10003060000fffffc002ffff7fff01ff000011c8001fffffc002ffff7fff01ff080407900007c0008003ffff7fff01ff480405b00007e1ff8003ffff7fff01ff90040d600003f1ff0003fffe7fff01ff10080fc00001f8fe0003fffe7fff00ff200923800000fe1e0003fffe7fff00ff200d03000000fffc0003fffe7fff00ff00011b000001fffc0003fffe7fff00ff0008d8000000fffc0003fffe7fff01ff0008780000003ff80003fffcfffe01ff300c3f0000001ff80003fffefff001ffd00e2d800003fffe00030ffeffe000ff800664c00007ffff8003c07effc000ff2002c5800007ffffe003f800ff80c0ff402357a00003fffff803fe04ff00e0ffc1408f100003fffffc03ff84fe00f8ff800107500007fffffe07ffedfc0078ffc00007a80005ffffff07fffdfc007cff800003d80043ffffff07fffff8003cff802001c80047ffffff87fffff8003cff802011980007ffffffc7ffeff8003cff800001c80007ffffffc7fffff0003cff8040099800c7ffffffc7fffff0001cff00000bf000c7ffffffe7ffdfe0001cff8000073800c7ffffffe7fffff0001cff8000038181c7ffffffe7ffffe0001cff800003c001cfffffffe7ffffc0001cff80000720018fffffffffffffc0001cff80000130038ffffffffcffffc0001cff80000800039ffffffff81fffc0001c7f80000800071ffffffff801ff80001e7f00000c00063ffffffff8007f00001e7f80000600063ffffffff8007f00001e7f800003400c7ffffffffc007e00001e7e8000008010fffffffffc007c00003e7e8000000000fffffffffc007800001e7f0000010003fffffffffc007000400e7f000201800ffffffffffe000000400e7f000101001ffffffffffe020000800f7e000100000ffffffffffe060801800f7f000080003fffffffffff071803000f7e000000007fffffffffff07f00300"`;
